url3426a97ee4_rtreffer_minidns_tgz-pJ8-de_measite_minidns_minidnsrepl_MiniDnsReplJ8/src/de/measite/minidns/DNSWorld.java

```diff
@@ -0,0 +1,549 @@
+
+        public Zone(String zoneName, InetAddress address, List<Record<? extends Data>> records) {
+            this.zoneName = zoneName;
+            this.address = address;
+            this.records = records;
+        }
+
+        public List<RRSet> getRRSets() {
+            List<RRSet.Builder> rrSetBuilders = new LinkedList<>();
+            outerloop: for (Record<? extends Data> record : records) {
+                for (RRSet.Builder builder : rrSetBuilders) {
+                    if (builder.addIfPossible(record)) {
+                        continue outerloop;
+                    }
+                }
+                rrSetBuilders.add(RRSet.builder().addRecord(record));
+            }
+            List<RRSet> rrSets = new ArrayList<>(rrSetBuilders.size());
+            for (RRSet.Builder builder : rrSetBuilders) {
+                rrSets.add(builder.build());
+            }
+            return rrSets;
+        }
+
+        boolean isRootZone() {
+            return (zoneName == null || zoneName.isEmpty()) && address == null;
+        }
+    }
+
+    public static DNSWorld applyZones(AbstractDNSClient client, Zone... zones) {
+        DNSWorld world = new DNSWorld();
+        client.setDataSource(world);
+        for (Zone zone : zones) {
+            for (RRSet rrSet : zone.getRRSets()) {
+                DNSMessage.Builder req = client.buildMessage(new Question(rrSet.name, rrSet.type, rrSet.clazz, false));
+                DNSMessage.Builder resp = DNSMessage.builder();
+                resp.setAnswers(rrSet.records);
+                resp.setAuthoritativeAnswer(true);
+                attachGlues(resp, rrSet.records, zone.records);
+                attachSignatures(resp, zone.records);
+                DNSMessage request = req.build();
+                DNSMessage response = resp.build();
+                if (zone.isRootZone()) {
+                    world.addPreparedResponse(new RootAnswerResponse(request, response));
+                } else {
+                    world.addPreparedResponse(new AddressedAnswerResponse(zone.address, request, response));
+                }
+                if (rrSet.type == TYPE.NS) {
+                    DNSMessage.Builder hintsResp = DNSMessage.builder();
+                    hintsResp.setNameserverRecords(rrSet.records);
+                    hintsResp.setAdditionalResourceRecords(response.additionalSection);
+                    DNSMessage hintsResponse = hintsResp.build();
+                    if (zone.isRootZone()) {
+                        world.addPreparedResponse(new RootHintsResponse(rrSet.name, hintsResponse));
+                    } else {
+                        world.addPreparedResponse(new AddressedHintsResponse(zone.address, rrSet.name, hintsResponse));
+                    }
+                }
+            }
+        }
+        return world;
+    }
+
+    static void attachSignatures(DNSMessage.Builder response, List<Record<? extends Data>> records) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 290-296, lines 301-307 into separate methods.